### PR TITLE
Add email confirmation on signup

### LINKED ISSUE #1183: Add an option to confirm email address on signup
Europe is launching a new law in May, called GDPR, which is great, but comes with a lot of constraints for anyone having a webservice.

One of the thing we need to do to be compliant is:

_With e-mail double opt-in, a new user first needs to verify their e-mail address before completing the sign-up process._

I've tried to prevent doing this on Monica since the beginning - not because I wanted to collect your email address and do nasty thing with it, but because I wanted people to sign up with fake email addresses if they wanted to.

This law will force us to confirm email addresses on signup.

We will create a new `.env` variable, called `SIGNUP_DOUBLE_OPTIN`, which will default to `TRUE`. On our hosted version, it will be set to true, but you will do whatever you want on your own instance.

Note: to be clear, we do have an email newsletter, but we only contact people who have subscribed manually and specifically to receive this newsletter. The email address used on signup is never used otherwise, other than sending reminder emails.

Changes to do:
- [ ] Create a email confirmation process
- [ ] Keep track of when the user has validated his email address
- [ ] Display this information in the Settings panel under a new section (called `Compliance`)
- [ ] Add a method in the User object, something like `isCompliant`, that will tell if users have confirmed emails and accepted Terms & Privacy. If not, we should block email sending (reminders, notifications,...) automatically. This will be done in https://github.com/monicahq/monica/issues/1184 so you don't have to do it here.

cc @asbiin @Mokto

### PR DESCRIPTION
**Pull request by asbiin**
This will close #1183 

#### Code-related tasks
- [ ] Tests added for this feature/bug.

#### If the code changes the SQL schema
- [X] Impact on account export.

#### Other tasks
- [X] [CHANGELOG](https://github.com/monicahq/monica/blob/master/CHANGELOG) entry added, if necessary, under `UNRELEASED`.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -59,6 +59,8 @@ APP_DEFAULT_LOCALE=en
 # Ability to disable signups on your instance.
 # Can be true or false. Default to false.
 APP_DISABLE_SIGNUP=true
+# Enable user email verification.
```
Comment (djaiss) [inline .env.example]: It's more elegant if you put an empty line above it.

File: app/Http/Requests/EmailChangeRequest.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'email' => 'required|string',
+            'password' => 'required|string',
+            'newmail' => 'required|email|max:2083|unique:users,email,'.$this->id,
```
Comment (djaiss) [inline app/Http/Requests/EmailChangeRequest.php]: Why `this->id` at the end?

File: app/Jobs/SendNewUserAlert.php [hunk 3]
```diff
@@ -33,6 +33,10 @@ public function __construct(User $user)
      */
     public function handle()
     {
-        Mail::to(config('monica.email_new_user_notification'))->send(new NewUserAlert($this->user));
+        $email = config('monica.email_new_user_notification');
```
Comment (djaiss) [inline app/Jobs/SendNewUserAlert.php]: 👍

File: app/Notifications/ConfirmEmail.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    /**
+     * Get the mail representation of the notification.
+     *
+     * @param  mixed  $notifiable
+     * @return \Illuminate\Notifications\Messages\MailMessage
+     */
+    public function toMail($notifiable)
+    {
+        return (new MailMessage)
+            ->subject(__('confirmation::confirmation.email-title'))
```
Comment (djaiss) [inline app/Notifications/ConfirmEmail.php]: What is `__('confirmation...` ?

Translations?

File: config/monica.php [hunk 5]
```diff
@@ -12,6 +12,16 @@
     */
     'disable_signup' => env('APP_DISABLE_SIGNUP', false),
 
+   /*
+    |--------------------------------------------------------------------------
+    | Disable User registration
```
Comment (djaiss) [inline config/monica.php]: Isn't it `Activate double optin on signup` instead?

File: app/Providers/RouteServiceProvider.php [hunk 6]
```diff
@@ -19,19 +19,25 @@
...
+
     /**
      * This namespace is applied to your controller routes.
      *
      * In addition, it is set as the URL generator's root namespace.
      *
      * @var string
      */
-    protected $namespace = 'App\Http\Controllers';
+    protected function getNamespace()
```
Comment (djaiss) [inline app/Providers/RouteServiceProvider.php]: I don't understand this part of having to change namespace. Can you explain?

File: .env.example (hunk 1 above)
Comment (asbiin) [reply .env.example]: OK

File: app/Http/Requests/EmailChangeRequest.php (hunk 2 above)
Comment (asbiin) [reply app/Http/Requests/EmailChangeRequest.php]: Indeed, I copied it from https://github.com/monicahq/monica/blob/master/app/Http/Requests/SettingsRequest.php#L29

File: app/Jobs/SendNewUserAlert.php (hunk 3 above)
Comment (asbiin) [reply app/Jobs/SendNewUserAlert.php]: 👍

File: app/Notifications/ConfirmEmail.php (hunk 4 above)
Comment (asbiin) [reply app/Notifications/ConfirmEmail.php]: Yes. This file is a stub copied from the new package.
I guess `__()` is an equivalent of `trans()`. The prefix 'confirmation::' is mandatory, it gets the translation from the package itself.

File: config/monica.php (hunk 5 above)
Comment (asbiin) [reply config/monica.php]: OK

File: app/Providers/RouteServiceProvider.php (hunk 6 above)
Comment (asbiin) [reply app/Providers/RouteServiceProvider.php]: It uses the trait `DetectsApplicationNamespace` from Laravel to get the root namespace, which I think is nice.

**Comment by asbiin** [comment]
image

File: app/Http/Controllers/Auth/EmailChangeController.php [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    /**
+     * Display a listing of the resource.
+     *
+     * @param  \Illuminate\Http\Request $request
+     * @return \Illuminate\Http\Response
+     */
+    public function index(Request $request)
+    {
+        if ($request->session()->has('user_id')) {
+            $model = config('auth.providers.users.model');
```
Comment (djaiss) [inline app/Http/Controllers/Auth/EmailChangeController.php]: I understand why you've done this, but in our case, I don't think we'll ever change the default `users.model` in an installation... so why do this?

File: config/app.php [hunk 8]
```diff
@@ -2,6 +2,19 @@
...
+    | Application Name
+    |--------------------------------------------------------------------------
+    |
+    | This value is the name of your application. This value is used when the
+    | framework needs to place the application's name in a notification or
+    | any other location as required by the application or its packages.
+    |
+    */
+
+    'name' => env('APP_NAME', 'Monica'),
```
Comment (djaiss) [inline config/app.php]: Holy cow. How come we didn't have this variable already?

File: resources/lang/en/auth.php [hunk 9]
```diff
@@ -59,4 +59,10 @@
     'register_policy' => 'Signing up signifies you’ve read and agree to our <a href=":url" hreflang=":hreflang">Privacy Policy</a> and <a href=":urlterm" hreflang=":hreflang">Terms of use</a>.',
     'register_invitation_email' => 'For security purposes, please indicate the email of the person who’ve invited you to join this account. This information is provided in the invitation email.',
 
+    'confirmation_again' => 'If you want to change your email you can <a href="'.url('settings/emailchange').'" class="alert-link">click here</a>.',
+    'email_change_current_email' => 'Current email address:',
+    'email_change_title' => 'Change your email',
+    'email_change_new' => 'New email address',
+    'email_change_password' => 'Confirm your password',
+    'email_changed' => 'Your email has been changed. Check your mailbox to validate it.',
```
Comment (djaiss) [inline resources/lang/en/auth.php]: `Your email address`

File: resources/lang/en/auth.php [hunk 10]
```diff
@@ -59,4 +59,10 @@
     'register_policy' => 'Signing up signifies you’ve read and agree to our <a href=":url" hreflang=":hreflang">Privacy Policy</a> and <a href=":urlterm" hreflang=":hreflang">Terms of use</a>.',
     'register_invitation_email' => 'For security purposes, please indicate the email of the person who’ve invited you to join this account. This information is provided in the invitation email.',
 
+    'confirmation_again' => 'If you want to change your email you can <a href="'.url('settings/emailchange').'" class="alert-link">click here</a>.',
```
Comment (djaiss) [inline resources/lang/en/auth.php]: `If you want to change your email address`

File: resources/lang/vendor/confirmation/en/confirmation.php [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+return [
+    'email-title' => 'Monica – Email verification',
+    'email-intro'=> 'To validate your email click on the button below',
+    'email-button' => 'Email verification',
```
Comment (djaiss) [inline resources/lang/vendor/confirmation/en/confirmation.php]: If it's a button it should be `Verify email address` or something.

File: app/Http/Controllers/Auth/EmailChangeController.php (hunk 7 above)
Comment (asbiin) [reply app/Http/Controllers/Auth/EmailChangeController.php]: Indeed. It's a copy/paste

File: config/app.php (hunk 8 above)
Comment (asbiin) [reply config/app.php]: Good question ...

**Comment by asbiin** [comment]
2fa problems are fixed. Everything should work now.
@djaiss can you review this a last time ?

**Pull request merged**